lib/switchtec.c

```diff
@@ -652,33 +698,306 @@ int switchtec_hard_reset(struct switchtec_dev *dev)
 			     NULL, 0);
 }
 
-static int log_a_to_file(struct switchtec_dev *dev, int sub_cmd_id, int fd)
+/**
+ * @brief Free log definition data
+ * @param[in] defs - log definition data to free
+ */
+static void free_log_defs(struct log_defs *defs)
+{
+	int i, j;
+
+	if (!defs->module_defs)
+		return;
+
+	for (i = 0; i < defs->num_alloc; i++) {
+		for (j = 0; j < defs->module_defs[i].num_entries; j++)
+			free(defs->module_defs[i].entries[j]);
+
+		free(defs->module_defs[i].entries);
+	}
+
+	free(defs->module_defs);
+}
+
+/**
+ * @brief Allocate / reallocate log definition data
+ * @param[in] defs 	  - log definition data
+ * @param[in] num_modules - number of modules to allocate for
+ * @return 0 on success, negative value on failure
+ */
+static int realloc_log_defs(struct log_defs *defs, int num_modules)
+{
+	int i;
+
+	defs->module_defs = realloc(defs->module_defs,
+				    (num_modules *
+				     sizeof(struct module_log_defs)));
+	if (!defs->module_defs) {
+		free_log_defs(defs);
+		return -1;
+	}
+
+	for (i = defs->num_alloc; i < num_modules; i++)
+		memset(&defs->module_defs[i], 0,
+		       sizeof(struct module_log_defs));
+
+	defs->num_alloc = num_modules;
+
+	return 0;
+}
+
+/**
+ * @brief Read a log definition file and store the definitions
+ * @param[in] log_def_file - log definition file
+ * @param[out] defs 	   - log definitions
+ * @return 0 on success, negative value on failure
+ */
+static int read_log_defs(FILE *log_def_file, struct log_defs *defs)
 {
 	int ret;
+	char line[LOG_ENTRY_LEN_MAX];
+	char mod_name[FW_MOD_NAME_LEN_MAX];
+	int mod_id;
+	int num_entries;
+	struct module_log_defs *mod_defs;
+	int i;
+
+	/* allocate some log definition entries */
+	ret = realloc_log_defs(defs, INIT_ALLOC_LOG_MODULES);
+	if (ret < 0)
+		return ret;
+
+	while (fgets(line, sizeof(line), log_def_file)) {
+		/*
+		 * Parse the module name, module ID, and number of module log
+		 * entries
+		 */
+		if (sscanf(line, "%31s %i %i",
```

Comment:
How about instead something like:

```
        tok = strtok(line, " \t");
        if (!tok)
            continue;
    
        tok = strtok(NULL, " \t");
        if (!tok)
            continue;
        mod_id = strtol(tok, NULL, 10);

        /* realloc, etc */

        tok = strtok(NULL, " \t");
        if (!tok)
            continue;

        mod_def->num_entries = strtol(tok, NULL, 10);
        mod_def->mod_name = strdup(line);
```

Removes the module name length mess.

Reply:
I did the sscanf since it's simpler and I know our module names won't be anywhere near 31 characters, but I'll use strtok instead just to be safe.